### PR TITLE
Ensure properties are dereferenced before comparing equality

### LINKED ISSUE #331: Error "incompatible properties" for properties compatible via $ref
The two `bar` properties should be considered compatible, because the types are effectively identical (one just flows through an extra `$ref`):

```
"FooBarString": {
  "type":"object",
  "properties": {
    "bar": {
      "type":"string"
    }
  },
  "allOf": [
    {
      "$ref": "#/definitions/FooBarStringRef"
    }
  ]
},
"FooBarStringRef": {
  "type":"object",
  "properties": {
    "bar": {
      "$ref":"#/definitions/MyString"
    }
  }
},
"MyString": {
  "type": "string"
}
```

Instead they generate an error:

```
    incompatible properties : bar
      definitions/FooBarString/properties/bar
        at file:///mnt/vss/_work/1/s/src/test/specs/compatible-properties-ref.json#L13:8
      definitions/FooBarStringRef/properties/bar
        at file:///mnt/vss/_work/1/s/src/test/specs/compatible-properties-ref.json#L26:8

```

### PR DESCRIPTION
**Pull request by mikeharder**
- Fixes #331

@johanste, @markcowl, @JeffreyRichter, @allenjzhang, @mikekistler: Please object if anything about this change looks **wrong** to you.

In summary, the change **always** calls `dereference()` on a property with a `$ref`, before comparing equality.

This looks pretty safe to me and @konrad-jamrozik, and passes the new unit test we wrote (plus all existing tests).

### CODE REVIEW

**Review by konrad-jamrozik — APPROVED** [review]

**Comment by konrad-jamrozik** [comment]
I see the test fails with `incompatible properties : bar`.

I took the spec:

``` json
{
  "swagger": "2.0",
  "info": {
    "title": "xms-enum-name",
    "version": "1.0"
  },
  "paths": {
  },
  "definitions": {
    "FooBarString": {
      "type":"object",
      "properties": {
        "bar": {
          "type":"string"
        }
      },
      "allOf": [
        {
          "$ref": "#/definitions/FooBarStringRef"
        }
      ]
    },
    "FooBarStringRef": {
      "type":"object",
      "properties": {
        "bar": {
          "$ref":"#/definitions/MyString"
        }
      }
    },
    "MyString": {
      "type": "string"
    }
  }
}
```

and pasted it here:

- https://editor.swagger.io/
- https://editor-next.swagger.io/

I got no errors in both cases.

image

Looking at that screenshot, both definitions have the same `bar`. There is no incompatibility anywhere. So the spec appears to be OK. The tool interpretation must be wrong.

**Comment by konrad-jamrozik** [comment]
OK I think I know what is the root-cuase. The problem lies in openapi-diff logic.

We throw error from [here](https://github.com/Azure/openapi-diff/blob/cb1ed36653f9b727a3190fad43fd6dfa861bf3ae/src/lib/util/resolveSwagger.ts#L259)

Which will happen if this line returns true:

`if (!this.isEqual(allOfSchema.properties[key], schemaList[key])) {`

On the failing test it evaluates to:

`this.isEqual({"$ref":"#/definitions/MyString"}, {"type":"string"})`

Now inside the `this.isEqual` logic the problem lies [here](https://github.com/Azure/openapi-diff/blob/cb1ed36653f9b727a3190fad43fd6dfa861bf3ae/src/lib/util/resolveSwagger.ts#L296):

`if ((!parentProperty.type || parentProperty.type === "object") && (!unwrappedProperty.type || unwrappedProperty.type === "object")) {`

On the failing test it evaluates to:

`(!undefined || undefined === "object") && (!string || string === "object")`

which is `(true && false)` which is `false`.

As a result, the `isEqual` will do no more dereferencing, so the `parentProperty`, which is `{"$ref":"#/definitions/MyString"}`, won't be dereferenced. Hence `isEqual` will return with the last `return`:

`return parentProperty.type === unwrappedProperty.type && parentProperty.format === unwrappedProperty.format`

which will evaluate like that:

`parentProperty.type (undefined) === unwrappedProperty.type (string) && parentProperty.format (undefined) === unwrappedProperty.format (undefined)`
which is
`undefined === string && undefined === undefined` which is false, hence we throw.

Basically, the logic of `isEqual` is assuming that it has either dereference both `parentProperty` and `unwrappedProperty`, or none, but not possibly only one.

**Comment by konrad-jamrozik** [comment]
@mikeharder I think you may want to change the PR title before merging to mention it fixes the bug.

File: src/lib/util/resolveSwagger.ts [hunk 1]
```diff
@@ -293,6 +293,23 @@ export class ResolveSwagger {
...
+    if (parentProperty.$ref || unwrappedProperty.$ref) {
+      return this.isEqual(
+        parentProperty.$ref ? this.dereference(parentProperty.$ref) : parentProperty,
+        unwrappedProperty.$ref ? this.dereference(unwrappedProperty.$ref) : unwrappedProperty
+      )
+    }
+
     if ((!parentProperty.type || parentProperty.type === "object") && (!unwrappedProperty.type || unwrappedProperty.type === "object")) {
       let parentPropertyToCompare = parentProperty
       let unwrappedPropertyToCompare = unwrappedProperty
```
Comment (mikeharder) [inline src/lib/util/resolveSwagger.ts:315]: I wonder if lines 316-321 are redundant now?  Since we (recursively) dereference both properties above.

Comment (konrad-jamrozik) [reply src/lib/util/resolveSwagger.ts:315]: They also do some object `===` equality.

Comment (konrad-jamrozik) [reply src/lib/util/resolveSwagger.ts:315]: Yeah probably. But I didn't want to deal with refactoring and increasing risk of breaking things.

Comment (konrad-jamrozik) [reply src/lib/util/resolveSwagger.ts:315]: But now when I think about it, it might be as simple as just deleting these lines.

File: src/lib/util/resolveSwagger.ts [hunk 2]
```diff
@@ -293,24 +293,26 @@ export class ResolveSwagger {
...
+
     if ((!parentProperty.type || parentProperty.type === "object") && (!unwrappedProperty.type || unwrappedProperty.type === "object")) {
-      let parentPropertyToCompare = parentProperty
-      let unwrappedPropertyToCompare = unwrappedProperty
-      if (parentProperty.$ref) {
-        parentPropertyToCompare = this.dereference(parentProperty.$ref)
-      }
-      if (unwrappedProperty.$ref) {
-        unwrappedPropertyToCompare = this.dereference(unwrappedProperty.$ref)
-      }
```
Comment (mikeharder) [inline src/lib/util/resolveSwagger.ts:304]: I think this is sufficient and simpler than the earlier recursive fix.  It's just moving lines 297-304 above the "if" check, and simplifying the code to update vars in-place.

File: src/lib/util/resolveSwagger.ts [hunk 3]
```diff
@@ -293,24 +293,26 @@ export class ResolveSwagger {
...
-    }
-    if (parentProperty.type === "array" && unwrappedProperty.type === "array") {
+      return parentProperty === unwrappedProperty
+    } else if (parentProperty.type === "array" && unwrappedProperty.type === "array") {
       return this.isEqual(parentProperty.items, unwrappedProperty.items)
+    } else {
+      return parentProperty.type === unwrappedProperty.type && parentProperty.format === unwrappedProperty.format
     }
-    return parentProperty.type === unwrappedProperty.type && parentProperty.format === unwrappedProperty.format
   }
```
Comment (mikeharder) [inline src/lib/util/resolveSwagger.ts:316]: This code should be equivalent, I just collapsed L305-308 into 310, and added `else`s for clarity.

File: src/lib/util/resolveSwagger.ts (hunk 1 above)
Comment (mikeharder) [reply src/lib/util/resolveSwagger.ts]: I took the risk.  LMK what you think.

**Review by konrad-jamrozik — APPROVED** [review]

**Review by mikekistler — APPROVED** [review]
Looks good! 👍

**Review by JeffreyRichter — APPROVED** [review]
